text/unification.md

```diff
@@ -0,0 +1,268 @@
+3. Interaction with type classes is not clear to me: shouldn't type classes inference be part of the loop trying heuristics so that:
+   - it is possible to mutually backtrack on choices made by heuristics and choices made by type classes inference (this lack of interaction happens to be a cause of the current unique failure in [#14805](https://github.com/coq/coq/issues/14805)).
+
+## About typing
+
+The first-order heuristic may produce ill-typed problems. PR [#14733](https://github.com/coq/coq/pull/14733) tries to ensure correct typing.
+
+When defining an existential variable with a term, the term is retyped. Is this really needed?
+
+# Directions for short terms improvements
+
+All the numbered points above are worth being addressed.
+
+My feeling is that the true backtracking ability C.1 of `Evarconv.solve_unif_constraints_with_heuristics` should come first followed by the extensions C.3 then C.2, so that the heuristics are strong enough to absorb/compensate incremental modifications in `evar_conv_x` and `solve_simple_eqn`.
+
+# Miscellaneous further perpectives
+
+**Delaying choices**
+- A way should be found to postpone the use of first-order unification heuristics as well as to postpone the decision of using canonical structures.
+  - For instance, for a `?P ?e =?= Q t` problem, one can set `?P := fun x => ?P'` with pending equation `?P'@{x:=?e} =?= Q t` and one could declare that `?e` has by default to be instantiated by `t` if no other instances is available.
+  - Similarly, for a `projᵢ ?e =?= t` problem, `?e` could be resolved into `Build ?e₁ ... t ... ?eₙ`, for `Build` the constructor of the record type, with some indication that each such subterm `Build ?e₁ ... t ... ?eₙ` with unsolved component should be resolved using a canonical equipment.
+
+**Extensions**
+- Extend canonical structures to canonical equipments based on knowing two or more projections?
+- Reorganize canonical structures so that it more generally solves equations of the form `proj₁ᵢ₁ (... projₙᵢₙ ?e ...) = u`. On one side, it should a priori give more discriminating solutions. On the other side, it would allow to treat canonical structures only in the flexible case.
+- Extend first-order unification to `match` term, e.g. to resolve `if true then t else u =?= if ?e₀ then ?e₁ else ?e₂` by setting `?e₀ := true`, `?e₁ := t` and `?e₂ := u` (see wish [#12780](https://github.com/coq/coq/issues/12780)).
+- Existential variables for sorts can sometimes not be instantiated by `Prop` even in situation where it would be legitimate to do so (see PR [#7369](https://github.com/coq/coq/pull/7369)).
+
+**Optimizations**
+- Systematically eta-reduce the solutions to existential variables (with all kinds of eta). On one side, this seems to be the natural expectation. On the other side, this would avoid caring about whether it might be dommageable to eta-expand this or that solution in the unification algorithm.
```

Comment:
```suggestion
- Systematically eta-reduce the solutions to existential variables (with all kinds of eta). On one side, this seems to be the natural expectation. On the other side, this would avoid caring about whether it might be harmful to eta-expand this or that solution in the unification algorithm.
```
Translating French to English with Google